modules/service/src/main/scala/lucuma/odb/service/ProgramUserService.scala

```diff
@@ -0,0 +1,495 @@
+// Copyright (c) 2016-2023 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.odb.service
+
+import cats.Monad
+import cats.data.NonEmptyList
+import cats.effect.Concurrent
+import cats.syntax.applicative.*
+import cats.syntax.applicativeError.*
+import cats.syntax.flatMap.*
+import cats.syntax.foldable.*
+import cats.syntax.functor.*
+import cats.syntax.functorFilter.*
+import cats.syntax.option.*
+import cats.syntax.traverse.*
+import grackle.Result
+import grackle.ResultT
+import grackle.syntax.*
+import lucuma.core.enums.Partner
+import lucuma.core.enums.ProgramUserRole
+import lucuma.core.model.Access
+import lucuma.core.model.GuestRole
+import lucuma.core.model.PartnerLink
+import lucuma.core.model.Program
+import lucuma.core.model.ServiceRole
+import lucuma.core.model.StandardRole
+import lucuma.core.model.User
+import lucuma.odb.data.OdbError
+import lucuma.odb.data.OdbErrorExtensions.*
+import lucuma.odb.data.UserType
+import lucuma.odb.graphql.input.AddProgramUserInput
+import lucuma.odb.graphql.input.ProgramUserPropertiesInput
+import lucuma.odb.graphql.input.UnlinkUserInput
+import lucuma.odb.util.Codecs.educational_status
+import lucuma.odb.util.Codecs.gender
+import lucuma.odb.util.Codecs.partner
+import lucuma.odb.util.Codecs.partner_link
+import lucuma.odb.util.Codecs.partner_link_type
+import lucuma.odb.util.Codecs.program_id
+import lucuma.odb.util.Codecs.program_user_role
+import lucuma.odb.util.Codecs.user_id
+import lucuma.odb.util.Codecs.user_type
+import lucuma.sso.client.SsoGraphQlClient
+import natchez.Trace
+import skunk.*
+import skunk.codec.boolean.bool
+import skunk.codec.text.varchar
+import skunk.data.Completion
+import skunk.syntax.all.*
+
+import Services.Syntax.*
+
+trait ProgramUserService[F[_]]:
+
+  /**
+   * Perform the requested program <-> user link, yielding the linked ids if successful, or None
+   * if the user was not authorized to perform the action.
+   */
+  def linkUser(
+    req: ProgramUserService.LinkUserRequest
+  )(using Transaction[F]): F[Result[(Program.Id, User.Id)]]
+
+  /** Unlink the requested user, yielding true if the user was unlinkes, false if no such link existed. */
+  def unlinkUser(
+    input: UnlinkUserInput
+  )(using Transaction[F], Services.PiAccess): F[Result[Boolean]]
+
+  def updateProgramUsers(
+    SET:   ProgramUserPropertiesInput,
+    which: AppliedFragment
+  )(using Transaction[F]): F[Result[List[(Program.Id, User.Id)]]]
+
+  def addProgramUser(
+    sso:   SsoGraphQlClient[F],
+    input: AddProgramUserInput
+  )(using NoTransaction[F]): F[Result[(Program.Id, User.Id)]]
```

Comment:
Most of this was just moved from `ProgramService` in order to group everything `ProgramUser` related.  This mutation though, `addProgramUser`, is new.